### PR TITLE
Disable ingress ip when cloud provider enabled

### PR DESCRIPTION
**Pull request by marun**
cc: @smarterclayton

### CODE REVIEW

**Comment by marun** [comment]
[test]

File: pkg/cmd/server/start/start_master.go [hunk 1]
```diff
@@ -676,11 +676,16 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 	oc.RunServiceServingCertController(serviceServingCertClient)
 	oc.RunUnidlingController()
 
-	_, _, ingressIPClient, err := oc.GetServiceAccountClients(bootstrappolicy.InfraServiceIngressIPControllerServiceAccountName)
-	if err != nil {
-		glog.Fatalf("Could not get client: %v", err)
+	onBareMetal := (kc.CloudProvider == nil)
+	if onBareMetal {
+		// The ingress ip controller should only be run on baremetal. In a cloud environment, it will conflict with the
```
Comment (smarterclayton) [inline pkg/cmd/server/start/start_master.go]: Have you tested whether you can run side by side even in a cloud environment?

Comment (marun) [inline pkg/cmd/server/start/start_master.go]: How would that be possible?  In a cloud environment the service controller is active and relies on the same status field that the ingress controller uses.

**Comment by smarterclayton** [comment]
There can be more than one ingress for a load balancer service.  Does the
cloud provider clobber them or merge?

**Comment by marun** [comment]
@smarterclayton It clobbers.  The assumption of the ingress ip controller is that it is in control of the status field.    This is the first time I'm hearing that the feature should work on a cloud provider.  Wasn't the whole point to provide an alternative on baremetal?

File: pkg/cmd/server/api/validation/master.go [hunk 2]
```diff
@@ -159,8 +159,14 @@ func ValidateMasterConfig(config *api.MasterConfig, fldPath *field.Path) Validat
...
-		if _, ipNet, err := net.ParseCIDR(cidr); err != nil || ipNet.IP.IsUnspecified() {
-			validationResults.AddErrors(field.Invalid(fldPath.Child("networkConfig", "ingressIPNetworkCIDR").Index(0), cidr, "must be a valid CIDR notation IP range (e.g. 172.30.0.0/16)"))
+		// TODO Detect cloud provider when not using built-in kubernetes
+		noCloudProvider := builtInKubernetes && (len(config.KubernetesMasterConfig.ControllerArguments["cloud-provider"]) == 0 || config.KubernetesMasterConfig.ControllerArguments["cloud-provider"][0] == "")
+		if noCloudProvider {
+			if _, ipNet, err := net.ParseCIDR(cidr); err != nil || ipNet.IP.IsUnspecified() {
+				validationResults.AddErrors(field.Invalid(fldPath.Child("networkConfig", "ingressIPNetworkCIDR").Index(0), cidr, "must be a valid CIDR notation IP range (e.g. 172.30.0.0/16)"))
+			}
+		} else {
+			validationResults.AddErrors(field.Invalid(fldPath.Child("networkConfig", "ingressIPNetworkCIDR").Index(0), cidr, "should not be provided when a cloud-provider is enabled."))
```
Comment (smarterclayton) [inline pkg/cmd/server/api/validation/master.go]: No period at the end.

File: pkg/cmd/server/api/validation/master.go [hunk 3]
```diff
@@ -159,8 +159,14 @@ func ValidateMasterConfig(config *api.MasterConfig, fldPath *field.Path) Validat
 	}
 	if len(config.NetworkConfig.IngressIPNetworkCIDR) > 0 {
 		cidr := config.NetworkConfig.IngressIPNetworkCIDR
-		if _, ipNet, err := net.ParseCIDR(cidr); err != nil || ipNet.IP.IsUnspecified() {
-			validationResults.AddErrors(field.Invalid(fldPath.Child("networkConfig", "ingressIPNetworkCIDR").Index(0), cidr, "must be a valid CIDR notation IP range (e.g. 172.30.0.0/16)"))
+		// TODO Detect cloud provider when not using built-in kubernetes
+		noCloudProvider := builtInKubernetes && (len(config.KubernetesMasterConfig.ControllerArguments["cloud-provider"]) == 0 || config.KubernetesMasterConfig.ControllerArguments["cloud-provider"][0] == "")
+		if noCloudProvider {
+			if _, ipNet, err := net.ParseCIDR(cidr); err != nil || ipNet.IP.IsUnspecified() {
+				validationResults.AddErrors(field.Invalid(fldPath.Child("networkConfig", "ingressIPNetworkCIDR").Index(0), cidr, "must be a valid CIDR notation IP range (e.g. 172.30.0.0/16)"))
```
Comment (smarterclayton) [inline pkg/cmd/server/api/validation/master.go]: Pick an example that corresponds with the default you intend to set

Comment (marun) [inline pkg/cmd/server/api/validation/master.go]: Addressed in #10500

Comment (smarterclayton) [inline pkg/cmd/server/api/validation/master.go]: Please add a test that exercises this in master_test.go

**Comment by knobunc** [comment]
The code (sans test) LGTM

**Comment by marun** [comment]
What a difference testing makes!

File: pkg/cmd/server/api/validation/master.go (hunk 3 above)
Comment (marun) [inline pkg/cmd/server/api/validation/master.go]: Done.

**Comment by smarterclayton** [comment]
Lgtm [merge]

**Comment by marun** [comment]
Check job failure is flake #10487

**Comment by marun** [comment]
Conformance job failure is flake #9548

**Pull request merged**